File: lang/swift/cxlib/Sources/CXLib/Table.swift
// CX Swift binding — Public Table API per ADR 0018 D1.
//
// 17-member canonical Table surface against the V core's `:table` blocks
// via the C ABI. Per ADR 0018 §D2: Swift uses camelCase methods matching
// the canonical surface.

import Foundation

/// Column iterator view (returned by `Table.iterCols()`).
public struct ColumnView {
    public let name: String
    public let typeName: String
    public let values: [Any?]
}

/// Immutable handle over a single `:table` block. Implements the
/// 17-member canonical Table API (ADR 0018 §D1).
public final class Table: Sequence {
    private let _cols: [String]
    private let _types: [String]
    private let _rows: [[Any?]]

    private init(cols: [String], types: [String], rows: [[Any?]]) {
        self._cols = cols
        self._types = types
        self._rows = rows
    }

    // ── Construction ─────────────────────────────────────────────────────────

    /// Parse CX source and return the first `:table` block.
    public static func fromCx(_ src: String) throws -> Table {
        let tables = try fromCxAll(src)
        guard let first = tables.first else {
            throw CXError.parse("cxlib: no :table block found in source")
        }
        return first
    }

    /// Return every `:table` block in the source (preorder).
    public static func fromCxAll(_ src: String) throws -> [Table] {
        let payload = try CXLib.toDataBin(src)
        let decoded = try DataBin.decode(payload)
        var out: [Table] = []
        collectTables(decoded, into: &out)
        return out
    }

    /// Direct construction with 4-invariant validation per ADR 0018 §D7.
    public static func create(
        cols: [String],
        types: [String],
        rows: [[Any?]]
    ) throws -> Table {
        guard cols.count == types.count else {
            throw CXError.parse("cxlib: len(cols)=\(cols.count) != len(types)=\(types.count)")
        }
        var seen = Set<String>()
        for c in cols {
            if !seen.insert(c).inserted {
                throw CXError.parse("cxlib: duplicate column name \"\(c)\"")
            }
        }
        for (i, r) in rows.enumerated() {
            if r.count != cols.count {
                throw CXError.parse("cxlib: row \(i) has \(r.count) cells; expected \(cols.count)")
            }
        }
        return Table(cols: cols, types: types, rows: rows)
    }

    // ── Properties (4) ───────────────────────────────────────────────────────

    public var cols: [String] { _cols }
    public var types: [String] { _types }
    public var rowCount: Int { _rows.count }
    public var colCount: Int { _cols.count }

    // ── Access (9) ───────────────────────────────────────────────────────────

    public func row(_ i: Int) throws -> [String: Any?] {
        guard i >= 0 && i < _rows.count else {
            throw CXError.parse("cxlib: row index \(i) out of bounds [0, \(_rows.count))")
        }
        var out: [String: Any?] = [:]
        for c in 0 ..< _cols.count { out[_cols[c]] = _rows[i][c] }
        return out
    }

    public func column(_ name: String) throws -> [Any?] {
        guard let idx = _cols.firstIndex(of: name) else {
            throw CXError.parse("cxlib: unknown column \"\(name)\"")
        }
        return try colAt(idx)
    }

    public func colAt(_ i: Int) throws -> [Any?] {
        guard i >= 0 && i < _cols.count else {
            throw CXError.parse("cxlib: column index \(i) out of bounds [0, \(_cols.count))")
        }
        return _rows.map { $0[i] }
    }

    public func cell(_ r: Int, _ c: Int) throws -> Any? {
        guard r >= 0 && r < _rows.count else {
            throw CXError.parse("cxlib: row index \(r) out of bounds [0, \(_rows.count))")
        }
        guard c >= 0 && c < _cols.count else {
            throw CXError.parse("cxlib: column index \(c) out of bounds [0, \(_cols.count))")
        }
        return _rows[r][c]
    }

    public func cellByName(_ r: Int, _ name: String) throws -> Any? {
        guard let idx = _cols.firstIndex(of: name) else {
            throw CXError.parse("cxlib: unknown column \"\(name)\"")
        }
        return try cell(r, idx)
    }

    public func slice(_ start: Int, _ end: Int) throws -> Table {
        guard start >= 0 && start <= _rows.count else {
            throw CXError.parse("cxlib: slice start \(start) out of bounds")
        }
        guard end >= start && end <= _rows.count else {
            throw CXError.parse("cxlib: slice end \(end) out of bounds (start=\(start))")
        }
        return Table(cols: _cols, types: _types, rows: Array(_rows[start ..< end]))
    }

    public func head(_ n: Int) throws -> Table {
        return try slice(0, Swift.min(Swift.max(n, 0), _rows.count))
    }

    public func tail(_ n: Int) throws -> Table {
        let start = Swift.max(0, _rows.count - n)
        return try slice(start, _rows.count)
    }

    /// `selectCols` — renamed from canonical `select` to mirror the
    /// other bindings' rename pattern.
    public func selectCols(_ names: [String]) throws -> Table {
        var indices: [Int] = []
        var newCols: [String] = []
        var newTypes: [String] = []
        for name in names {
            guard let idx = _cols.firstIndex(of: name) else {
                throw CXError.parse("cxlib: unknown column \"\(name)\"")
            }
            indices.append(idx)
            newCols.append(_cols[idx])
            newTypes.append(_types[idx])
        }
        let newRows = _rows.map { row in indices.map { row[$0] } }
        return Table(cols: newCols, types: newTypes, rows: newRows)
    }

    // ── Iteration (2) ────────────────────────────────────────────────────────

    public func makeIterator() -> AnyIterator<[String: Any?]> {
        var i = 0
        return AnyIterator {
            guard i < self._rows.count else { return nil }
            defer { i += 1 }
            return try? self.row(i)
        }
    }

    public func iterCols() -> [ColumnView] {
        var out: [ColumnView] = []
        for i in 0 ..< _cols.count {
            out.append(ColumnView(
                name: _cols[i],
                typeName: _types[i],
                values: (try? colAt(i)) ?? []
            ))
        }
        return out
    }

    // ── Conversion (5) ───────────────────────────────────────────────────────

    public func toCx() -> String {
        var s = "[_ :table["
        for i in 0 ..< _cols.count {
            if i > 0 { s += " " }
            s += _cols[i]
            if !_types[i].isEmpty { s += ":" + _types[i] }
        }
        s += "]\n"
        for row in _rows {
            s += "  "
            for (i, v) in row.enumerated() {
                if i > 0 { s += " " }
                s += formatCxCell(v)
            }
            s += "\n"
        }
        s += "]\n"
        return s
    }

    public func toCsv(delim: Character = ",") -> String {
        var s = ""
        for (i, c) in _cols.enumerated() {
            if i > 0 { s.append(delim) }
            s += c
        }
        s += "\r\n"
        for row in _rows {
            for (i, v) in row.enumerated() {
                if i > 0 { s.append(delim) }
                s += formatCsvCell(v, delim: delim)
            }
            s += "\r\n"
        }
        return s
    }

    public func toJson() throws -> String {
        let list = toDictList()
        let normalized = list.map { dict -> [String: Any] in
            var out: [String: Any] = [:]
            for (k, v) in dict { out[k] = v ?? NSNull() }
            return out
        }
        let data = try JSONSerialization.data(withJSONObject: normalized, options: [.sortedKeys])
        return String(decoding: data, as: UTF8.self)
    }

    public func toDataBin() throws -> Data {
        return try CXLib.toDataBin(toCx())
    }

    public func toDictList() -> [[String: Any?]] {
        var out: [[String: Any?]] = []
        for i in 0 ..< _rows.count {
            if let r = try? row(i) { out.append(r) }
        }
        return out
    }

    // ── Equality ─────────────────────────────────────────────────────────────

    public func equals(_ other: Table) -> Bool {
        if _cols != other._cols { return false }
        if _types != other._types { return false }
        if _rows.count != other._rows.count { return false }
        for r in 0 ..< _rows.count {
            if _rows[r].count != other._rows[r].count { return false }
            for c in 0 ..< _rows[r].count {
                if !cellEquals(_rows[r][c], other._rows[r][c]) { return false }
            }
        }
        return true
    }
}

// ── Internal: walk decoded data_bin value to find tables ────────────────────

private func collectTables(_ value: Any?, into out: inout [Table]) {
    if value == nil { return }
    if let dict = value as? [String: Any] {
        for child in dict.values { collectTables(child, into: &out) }
        return
    }
    if let list = value as? [Any] {
        if looksLikeTable(list) {
            let rows = list.compactMap { $0 as? [String: Any] }
            let keys = rows[0].keys.sorted()
            let types = Array(repeating: "", count: keys.count)
            let rowVals = rows.map { d -> [Any?] in
                keys.map { k -> Any? in
                    if let v = d[k] {
                        if v is NSNull { return nil }
                        return v
                    }
                    return nil
                }
            }
            if let t = try? Table.create(cols: keys, types: types, rows: rowVals) {
                out.append(t)
            }
            return
        }
        for child in list { collectTables(child, into: &out) }
    }
}

private func looksLikeTable(_ list: [Any]) -> Bool {
    if list.isEmpty { return false }
    guard let first = list[0] as? [String: Any] else { return false }
    let keys = Set(first.keys)
    if keys.isEmpty { return false }
    for item in list {
        guard let d = item as? [String: Any] else { return false }
        if d.count != keys.count { return false }
        for k in d.keys { if !keys.contains(k) { return false } }
    }
    return true
}

// ── Internal: cell formatters / equality ────────────────────────────────────

private func formatCxCell(_ v: Any?) -> String {
    guard let v = v else { return "null" }
    if let b = v as? Bool { return b ? "true" : "false" }
    if let i = v as? Int { return String(i) }
    if let i = v as? Int64 { return String(i) }
    if let d = v as? Double { return String(d) }
    if let f = v as? Float { return String(f) }
    if let list = v as? [Any?] {
        return "[" + list.map { formatCxCell($0) }.joined(separator: ", ") + "]"
    }
    if let list = v as? [Any] {
        return "[" + list.map { formatCxCell($0) }.joined(separator: ", ") + "]"
    }
    if let dict = v as? [String: Any?] {
        let keys = dict.keys.sorted()
        return "{" + keys.map { "\(formatCxKey($0)): \(formatCxCell(dict[$0] ?? nil))" }.joined(separator: ", ") + "}"
    }
    if let dict = v as? [String: Any] {
        let keys = dict.keys.sorted()
        return "{" + keys.map { "\(formatCxKey($0)): \(formatCxCell(dict[$0]))" }.joined(separator: ", ") + "}"
    }
    let s = "\(v)"
    if s.isEmpty || s.contains(where: { $0.isWhitespace || "'[](){},".contains($0) }) {
        return "'" + s.replacingOccurrences(of: "'", with: "''") + "'"
    }
    return s
}

private func formatCxKey(_ k: String) -> String {
    var simple = !k.isEmpty
    for (i, ch) in k.enumerated() {
        let ok = ch == "_" || ch.isLetter || (i > 0 && ch.isNumber)
        if !ok { simple = false; break }
    }
    if simple { return k }
    return "'" + k.replacingOccurrences(of: "'", with: "''") + "'"
}

private func formatCsvCell(_ v: Any?, delim: Character) -> String {
    guard let v = v else { return "" }
    if v is [Any] || v is [String: Any] || v is [Any?] || v is [String: Any?] {
        let data = (try? JSONSerialization.data(withJSONObject: v, options: [])) ?? Data()
        let s = String(decoding: data, as: UTF8.self)
        return "\"" + s.replacingOccurrences(of: "\"", with: "\"\"") + "\""
    }
    if let b = v as? Bool { return b ? "true" : "false" }
    let s = "\(v)"
    if s.contains(delim) || s.contains("\"") || s.contains("\n") || s.contains("\r") {
        return "\"" + s.replacingOccurrences(of: "\"", with: "\"\"") + "\""
    }
    return s
}

private func cellEquals(_ a: Any?, _ b: Any?) -> Bool {
    if a == nil && b == nil { return true }
    guard let a = a, let b = b else { return false }
    if let la = a as? [Any?], let lb = b as? [Any?] {
        if la.count != lb.count { return false }
        for i in 0 ..< la.count { if !cellEquals(la[i], lb[i]) { return false } }
        return true
    }
    if let la = a as? [Any], let lb = b as? [Any] {
        if la.count != lb.count { return false }
        for i in 0 ..< la.count { if !cellEquals(la[i], lb[i]) { return false } }
        return true
    }
    if let da = a as? [String: Any], let db = b as? [String: Any] {
        if da.count != db.count { return false }
        for (k, v) in da { if !cellEquals(v, db[k]) { return false } }
        return true
    }
    if let ai = a as? Int, let bi = b as? Int { return ai == bi }
    if let al = a as? Int64, let bl = b as? Int64 { return al == bl }
    if let ad = a as? Double, let bd = b as? Double { return ad == bd }
    if let ab = a as? Bool, let bb = b as? Bool { return ab == bb }
    if let astr = a as? String, let bstr = b as? String { return astr == bstr }
    return false
}

File: lang/ruby/cxlib/lib/cxlib/table.rb
# frozen_string_literal: true

# CX Ruby binding — Public Table API per ADR 0018 D1.
#
# 17-member canonical Table surface against the V core's :table blocks
# via the C ABI. Per ADR 0018 §D2: Ruby uses snake_case methods matching
# the canonical surface (rename `select` → `select_cols` to avoid
# clobbering Enumerable#select via include).

require 'json'

module CXLib
  class Table
    include Enumerable

    # ── Construction ───────────────────────────────────────────────────────

    # Parse CX source and return the first :table block.
    def self.from_cx(src)
      tables = from_cx_all(src)
      raise "cxlib: no :table block found in source" if tables.empty?
      tables.first
    end

    # Return every :table block in the source (preorder).
    def self.from_cx_all(src)
      payload = CXLib.to_data_bin(src)
      decoded = CXLib::DataBin.decode(payload)
      out = []
      _collect_tables(decoded, out)
      out
    end

    # Direct construction with 4-invariant validation per ADR 0018 §D7.
    def self.create(cols, types, rows)
      raise "cxlib: len(cols)=#{cols.size} != len(types)=#{types.size}" if cols.size != types.size
      seen = {}
      cols.each do |c|
        raise "cxlib: duplicate column name \"#{c}\"" if seen.key?(c)
        seen[c] = true
      end
      rows.each_with_index do |row, i|
        raise "cxlib: row #{i} has #{row.size} cells; expected #{cols.size}" if row.size != cols.size
      end
      allocate.tap { |t| t.send(:_init, cols.dup, types.dup, rows.map(&:dup)) }
    end

    def initialize(cols, types, rows)
      _init(cols, types, rows)
    end

    # ── Properties (4) ─────────────────────────────────────────────────────

    def cols      = @cols.dup
    def types     = @types.dup
    def row_count = @rows.size
    def col_count = @cols.size

    # ── Access (9) ─────────────────────────────────────────────────────────

    def row(i)
      raise "cxlib: row index #{i} out of bounds [0, #{@rows.size})" unless i.between?(0, @rows.size - 1)
      h = {}
      @cols.each_with_index { |name, c| h[name] = @rows[i][c] }
      h
    end

    def column(name)
      idx = @cols.index(name)
      raise "cxlib: unknown column \"#{name}\"" unless idx
      col_at(idx)
    end

    def col_at(i)
      raise "cxlib: column index #{i} out of bounds [0, #{@cols.size})" unless i.between?(0, @cols.size - 1)
      @rows.map { |row| row[i] }
    end

    def cell(r, c)
      raise "cxlib: row index #{r} out of bounds [0, #{@rows.size})" unless r.between?(0, @rows.size - 1)
      raise "cxlib: column index #{c} out of bounds [0, #{@cols.size})" unless c.between?(0, @cols.size - 1)
      @rows[r][c]
    end

    def cell_by_name(r, name)
      idx = @cols.index(name)
      raise "cxlib: unknown column \"#{name}\"" unless idx
      cell(r, idx)
    end

    def slice(start_idx, end_idx)
      raise "cxlib: slice start #{start_idx} out of bounds" unless start_idx.between?(0, @rows.size)
      raise "cxlib: slice end #{end_idx} out of bounds (start=#{start_idx})" unless end_idx.between?(start_idx, @rows.size)
      Table.create(@cols, @types, @rows[start_idx...end_idx].map(&:dup))
    end

    def head(n)
      slice(0, [[n, 0].max, @rows.size].min)
    end

    def tail(n)
      start_idx = [0, @rows.size - n].max
      slice(start_idx, @rows.size)
    end

    # `select_cols` — renamed from canonical `select` (Ruby's
    # Enumerable#select via `include` would clobber it).
    def select_cols(names)
      indices = names.map do |n|
        idx = @cols.index(n)
        raise "cxlib: unknown column \"#{n}\"" unless idx
        idx
      end
      new_cols  = indices.map { |i| @cols[i] }
      new_types = indices.map { |i| @types[i] }
      new_rows  = @rows.map { |row| indices.map { |i| row[i] } }
      Table.create(new_cols, new_types, new_rows)
    end

    # ── Iteration (2) ──────────────────────────────────────────────────────

    def each
      return enum_for(:each) unless block_given?
      (0...@rows.size).each { |i| yield row(i) }
    end

    def iter_cols
      return enum_for(:iter_cols) unless block_given?
      @cols.each_with_index do |name, i|
        yield ColumnView.new(name, @types[i], col_at(i))
      end
    end

    # ── Conversion (5) ─────────────────────────────────────────────────────

    def to_cx
      header = @cols.each_with_index.map { |c, i| @types[i].empty? ? c : "#{c}:#{@types[i]}" }.join(' ')
      lines = ["[_ :table[#{header}]"]
      @rows.each do |row|
        lines << '  ' + row.map { |v| Table._format_cx_cell(v) }.join(' ')
      end
      lines << ']'
      lines.join("\n") + "\n"
    end

    def to_csv(delim: ',')
      raise "cxlib: to_csv delim must be 1 char; got #{delim.size}" if delim.size != 1
      lines = [@cols.join(delim)]
      @rows.each do |row|
        lines << row.map { |v| Table._format_csv_cell(v, delim) }.join(delim)
      end
      lines.join("\r\n") + "\r\n"
    end

    def to_json(*_args)
      JSON.generate(to_dict_list)
    end

    def to_data_bin
      CXLib.to_data_bin(to_cx)
    end

    def to_dict_list
      (0...@rows.size).map { |i| row(i) }
    end

    # ── Equality ───────────────────────────────────────────────────────────

    def ==(other)
      return false unless other.is_a?(Table)
      @cols == other.cols && @types == other.types && @rows == other.instance_variable_get(:@rows)
    end
    alias eql? ==

    def hash
      [@cols, @types, @rows.size].hash
    end

    # ── Internal: walk decoded data_bin value to find tables ───────────────

    def self._collect_tables(value, out)
      case value
      when nil
        nil
      when Hash
        value.each_value { |child| _collect_tables(child, out) }
      when Array
        if _looks_like_table(value)
          first = value.first
          keys = first.keys.sort
          types = keys.map { '' }
          rows = value.map { |item| keys.map { |k| item[k] } }
          out << create(keys, types, rows)
        else
          value.each { |child| _collect_tables(child, out) }
        end
      end
    end

    def self._looks_like_table(arr)
      return false if arr.empty?
      first = arr.first
      return false unless first.is_a?(Hash)
      keys = first.keys
      return false if keys.empty?
      key_set = keys.to_set
      arr.all? do |item|
        item.is_a?(Hash) && item.size == keys.size && item.keys.all? { |k| key_set.include?(k) }
      end
    end

    def self._format_cx_cell(v)
      case v
      when nil then 'null'
      when true then 'true'
      when false then 'false'
      when Integer, Float then v.to_s
      when Array then '[' + v.map { |x| _format_cx_cell(x) }.join(', ') + ']'
      when Hash
        keys = v.keys.sort
        '{' + keys.map { |k| "#{_format_cx_key(k)}: #{_format_cx_cell(v[k])}" }.join(', ') + '}'
      else
        s = v.to_s
        if s.empty? || s =~ /[\s'\[\](){},]/
          "'" + s.gsub("'", "''") + "'"
        else
          s
        end
      end
    end

    def self._format_cx_key(k)
      return k if k =~ /\A[A-Za-z_][A-Za-z0-9_]*\z/
      "'" + k.gsub("'", "''") + "'"
    end

    def self._format_csv_cell(v, delim)
      return '' if v.nil?
      return '"' + JSON.generate(v).gsub('"', '""') + '"' if v.is_a?(Hash) || v.is_a?(Array)
      return v ? 'true' : 'false' if v == true || v == false
      s = v.to_s
      if s.include?(delim) || s.include?('"') || s.include?("\n") || s.include?("\r")
        '"' + s.gsub('"', '""') + '"'
      else
        s
      end
    end

    private

    def _init(cols, types, rows)
      @cols = cols
      @types = types
      @rows = rows
    end
  end

  # Column iterator view.
  ColumnView = Struct.new(:name, :type_name, :values)
end

require 'set'
